### DISCUSSION TITLE
The root_path is missing from the redirects from an url with trailing slash to the same url without it

### DISCUSSION THREAD

**Discussion by Mattiam2** [fastapi/fastapi#11153] [Questions]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import FastAPI

app = FastAPI(root_path="/wms")


@app.get("/hello")
async def root():
    return {"message": "Hello World"}
```

### Description

The program runs in gunicorn (using the uvicorn worker class) behind a nginx proxy which strips away the root_path (as described in [Behind a proxy](https://fastapi.tiangolo.com/advanced/behind-a-proxy/#setting-the-root_path-in-the-fastapi-app)).

Before fastapi 0.109 an URL with a trailing slash (ex. https://127.0.0.1/wms/hello/) would be redirected to the corresponding URL without the trailing slash (ex. https://127.0.0.1/wms/hello) as expected.

From fastapi 0.109 the redirect no more includes the root_path, so from https://127.0.0.1/wms/hello/ the application redirects to https://127.0.0.1/hello which results in a 404 Not Found error.

Here the headers of the redirect request:
![image](https://github.com/tiangolo/fastapi/assets/1143780/2782cab2-269e-4b84-b6da-a313b372629b)

I think the issue is correlated to [this one](https://github.com/tiangolo/fastapi/issues/10978) but it's not really the same.

### Operating System

Linux

### Operating System Details

The server runs on Redhat 8.7

### FastAPI Version

0.109.2

### Pydantic Version

2.6.1

### Python Version

3.9.13

**Comment by Kludex** [comment]
What's the uvicorn version? If it's not the latest one, try it.

**Reply by Mattiam2** [reply]
I'm using uvicorn 0.27.1 which is the latest one

**Comment by Mattiam2** [comment]
I think I've found where the issue came from.
In starlette 0.35.0 it was changed the constructor of the URL data structure removing the root_path from the path variable (you can see the change [here](https://github.com/encode/starlette/commit/c3c6314759b8356df185857c6001de6c8575708d#diff-e939ae940ce35a0db1544bfea4ab853be43ba89ea33e3cac713aa0b19f4bb875)).
Is this the right behaviour? It doesn't seem right to me, by the way I'll open an issue on Starlette.

**Reply by dietmar** [reply]
Have you found an answer? Did you end up creating a Starlette issue and could you provide the link? Thanks!

**Reply by Mattiam2** [reply]
Yes, I've opened an issue here: https://github.com/encode/starlette/discussions/2514

In the meanwhile because I use a reverse proxy (nginx), I've managed to "fix" the issue by specifying the root_path when using proxy_pass with the socket.

Before:
```
location ~ ^/wms/(.*)$ {
    proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
    proxy_set_header Host $http_host;
    proxy_redirect off;

    if (!-f $request_filename) {
        proxy_pass http://unix:/run/gunicorn/wms.sock:/$1$is_args$args;
        break;
    }
}
```

After:
```
location ~ ^/wms/(.*)$ {
    proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
    proxy_set_header Host $http_host;
    proxy_redirect off;

    if (!-f $request_filename) {
        proxy_pass http://unix:/run/gunicorn/wms.sock:/wms/$1$is_args$args;
        break;
    }
}
```

Or by just removing the arguments all together (in this way the full path will be used):
`proxy_pass http://unix:/run/gunicorn/wms.sock;`

This solution works enough for me but shouldn't be correct, because for the way root_path is defined within fastapi/starlette, fastapi socket shouldn't expect to have the root_path included in the url path.

**Comment by ajsmith-savills** [comment]
I can confirm I am also seeing this issue using fastapi v0.115.0 with starlette v0.38.6. I a building an API on a subsite and want the root request to auto resdirect to the docs page
host/subsite -> host/subsite/docs
Instead I am seeing
host/subsite -> host/docs

**Reply by ajsmith-savills** [reply]
In fact in localdev I am seeing the port get dropped too
127.0.0.1:8000/subsite -> 127.0.0.1/docs

**Comment by YuriiMotov** [comment]
The thing here is that you shouldn't strip path prefix when you specify `root_path` as a parameter to `FastAPI` (`FastAPI(root_path="/wms")`).

According to ASGI spec, `path` in scope MUST include prefix.
When you run app with Uvicorn with `--root-path` option, Uvicorn adds this prefix back. So, proxy strips it, but Uvicorn adds it back.
In your configuration, Uvicron (or whatever server you are using) doesn't add it, and you should configure proxy not to strip it.

IMO, this is a mistake in docs and I attempt to fix it in this PR: https://github.com/fastapi/fastapi/pull/14238
